oak/server/rust/oak_runtime/src/node/grpc_server.rs

```diff
@@ -0,0 +1,334 @@
+            self.runtime.channel_create(&Label::public_trusted());
+
+        // Create an invocation message and attach the method-invocation specific channels to it.
+        let invocation = crate::Message {
+            data: vec![],
+            channels: vec![request_reader, response_writer],
+        };
+
+        // Serialize gRPC request into a message.
+        let mut message = crate::Message {
+            data: vec![],
+            channels: vec![],
+        };
+        request
+            .write_to_writer(&mut message.data)
+            .map_err(|error| {
+                error!("Couldn't serialize a GrpcRequest message: {}", error);
+                GrpcServerError::RequestProcessingError
+            })?;
+
+        // Send a message to the temporary channel.
+        self.runtime
+            .channel_write(request_writer, message)
+            .map_err(|error| {
+                error!(
+                    "Couldn't write a message to the terporary channel: {:?}",
+                    error
+                );
+                GrpcServerError::RequestProcessingError
+            })?;
+
+        // Send an invocation message (with attached handles) to the Oak node.
+        self.runtime
+            .channel_write(
+                self.node_writer.expect("Node writer wasn't initialized"),
+                invocation,
+            )
+            .map_err(|error| {
+                error!("Couldn't write a gRPC invocation message: {:?}", error);
+                GrpcServerError::RequestProcessingError
+            })?;
+
+        Ok(response_reader)
+    }
+
+    /// Processes a gRPC response from a channel represented by `response_reader` and returns an
+    /// HTTP response body.
+    fn process_response(&self, response_reader: Handle) -> Result<Vec<u8>, GrpcServerError> {
+        let read_status = self
+            .runtime
+            .wait_on_channels(&[Some(response_reader)])
+            .map_err(|error| {
+                error!("Couldn't wait on the temporary gRPC channel: {:?}", error);
+                GrpcServerError::ResponseProcessingError
+            })?;
+
+        if read_status[0] == ChannelReadStatus::ReadReady {
+            self.runtime
+                .channel_read(response_reader)
+                .map_err(|error| {
+                    error!("Couldn't read temporary gRPC channel: {:?}", error);
+                    GrpcServerError::ResponseProcessingError
+                })
+                .map(|message| {
+                    // Return an empty HTTP body if the `message` is None.
+                    message.map_or(vec![], |m| m.data)
+                })
+        } else {
+            error!("Couldn't read channel: {:?}", read_status[0]);
+            Err(GrpcServerError::ResponseProcessingError)
+        }
+    }
+}
+
+/// Oak Node implementation for the gRPC server.
+impl super::Node for GrpcServerNode {
+    fn start(&mut self) -> Result<(), OakStatus> {
```

Comment:
The gRPC seems to be combining two different Hyper concepts into one: The server, which is listening for new requests and the service, which handles each request. I think this is why there are so many clones statements needed.

Reply:
This logic was merged into the Node based on this thread: https://github.com/project-oak/oak/pull/772#discussion_r401123844
But even if we will have a separate `service` object, we'll still need to clone it, since it should live inside a Future.

Reply:
As part of the cleanup in #813, it might be worth looking at using a [Tower service](https://github.com/hyperium/hyper/blob/master/examples/tower_server.rs#L14) (implemented as a gRPC service node), which could encapsulate the needed state so that it does not have to be cloned so many times. It would also reduce the need for so many nested `move`s and `async move`s.

Reply:
They are necessary in the Node thread, but not in the Future, yes